### ISSUE TITLE
Building using MinGW-w64 GCC in Windows (MINGW64 environment in MSYS2)

### ISSUE THREAD

**Issue by przemoc** [wolfpld/tracy#252]
# Intro

I stumbled upon Tracy Profiler via @andrewrk video ([Sub-Millisecond Compilations with Zig’s In-Place Binary Patching](https://vimeo.com/491488902)) linked from @kristoff-it post ([Zig's New Relationship with LLVM](https://kristoff.it/blog/zig-new-relationship-llvm/)). Looks sleek, so I wanted to give it a try yesterday, it didn't go smoothly, so wanted to leave some notes, as they may be potentially helpful for others.

# Environment

    HW: ThinkPad T430 (Intel(R) Core(TM) i7-3520M CPU @ 2.90 GHz)
    OS: Windows 7 Pro (6.1.7601 Service Pack 1)

On Windows my build platform of choice is [MSYS2](https://www.msys2.org/), typically with [MinGW-w64](http://mingw-w64.org/), started with `MSYS2 MinGW 64-bit` shortcut (`msys2_shell.cmd -mingw64`), which I'm keeping up-to-date:

    $ pacman -Q mingw-w64-x86_64-gcc
    mingw-w64-x86_64-gcc 10.3.0-5
    $ gcc --version | head -1 && gcc -dumpmachine
    gcc.exe (Rev5, Built by MSYS2 project) 10.3.0
    x86_64-w64-mingw32

# Preparations

After cloning master (wolfpld/tracy@e380e3f80e45 at the time), I installed some packages:

    pacman -S mingw-w64-x86_64-freetype mingw-w64-x86_64-capstone mingw-w64-x86_64-glfw mingw-w64-x86_64-gtk3

so I could build the server. I went to `profiler/build/unix/` and did:

    make -j3

There were a few issues along the way, but after some trials and errors, googling, etc., I managed to build it (final successful build took 45 minutes, but take it with grain of salt, as my laptop has fan issues, so CPU is getting throttled easily). I didn't save the error messages at various stages, but let me comment what I had to do and summary of all these extra steps  at the end.

# Issues

1. To have [`dlfcn.h`](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/dlfcn.h.html) POSIX stuff (+ [`dladdr()`](https://man7.org/linux/man-pages/man3/dladdr.3.html) which is not yet part of POSIX, but [will be](https://www.austingroupbugs.net/view.php?id=993)) in Windows we need a library that wraps Windows equivalent functions into POSIX-compatible API. There exists a project doing exactly this: [dlfcn-win32/dlfcn-win32](https://github.com/dlfcn-win32/dlfcn-win32).

2. To fix issues about undefined `WSAStartup()`, `WSAPoll()`, etc., we have to link with `ws2_32`.

3. To deal with `File too big` and `too many sections (...)`, we have to pass `-mbig-obj` flag to assembler.

4. To make LTO build successful, we have to add following compilation flags `-fuse-linker-plugin -ffat-lto-objects`.

5. To fix Tracy to not hardcode `TracyLzcnt` into `__lzcnt64`, as `LZCNT` instruction (Count Leading Zeros) is not available before Haswell aka 4th gen core, like in my Ivy Bridge aka 3th gen core (`error: inlining failed in call to 'always_inline' '__lzcnt64': target specific option mismatch`), but built-in intrinsic function instead, we need to ensure that despite `_WIN64` set, hardcoding won't be done if `__MINGW32__` is defined, so that other preprocessor branch will be taken (`#elif defined __GNUC__ || defined __clang__`, where `__builtin_clzll` is used).

# tl;dr

Install dlfcn:

    pacman -S mingw-w64-x86_64-dlfcn

Apply following changes in tracy repository:

``` diff
diff --git a/profiler/build/unix/build.mk b/profiler/build/unix/build.mk
index f403b879c6f7..913c208b0d00 100644
--- a/profiler/build/unix/build.mk
+++ b/profiler/build/unix/build.mk
@@ -2,7 +2,7 @@ CFLAGS +=
 CXXFLAGS := $(CFLAGS) -std=c++17
 DEFINES += -DIMGUI_IMPL_OPENGL_LOADER_GL3W -DIMGUI_ENABLE_FREETYPE
 INCLUDES := $(shell pkg-config --cflags glfw3 freetype2 capstone) -I../../../imgui -I../../libs/gl3w
-LIBS := $(shell pkg-config --libs glfw3 freetype2 capstone) -lpthread -ldl
+LIBS := $(shell pkg-config --libs glfw3 freetype2 capstone) -lpthread -ldl -lws2_32

 DISPLAY_SERVER := X11

diff --git a/profiler/build/unix/release.mk b/profiler/build/unix/release.mk
index e70f36d486ad..346fcd3d4635 100644
--- a/profiler/build/unix/release.mk
+++ b/profiler/build/unix/release.mk
@@ -1,4 +1,4 @@
-CFLAGS := -O3 -flto
+CFLAGS := -O3 -flto -Wa,-mbig-obj -fuse-linker-plugin -ffat-lto-objects
 DEFINES := -DNDEBUG
 BUILD := release

diff --git a/server/TracyPopcnt.hpp b/server/TracyPopcnt.hpp
index 945324fd91d6..4a29c4f8de72 100644
--- a/server/TracyPopcnt.hpp
+++ b/server/TracyPopcnt.hpp
@@ -4,7 +4,7 @@
 #include <limits.h>
 #include <stdint.h>

-#if defined _WIN64
+#if defined _WIN64 && !defined __MINGW32__
 #  include <intrin.h>
 #  define TracyCountBits __popcnt64
 #  define TracyLzcnt __lzcnt64
```

# Disclaimer

I managed to built Tracy Profiler server, and it starts executing without issues (shows Tracy Profiler 0.7.9 windows with Get started dialog box), but I did run out of time and did not manage to confirm if it's working as it should (I have no experience with Tracy Profiler whatsoever), so I'm not claiming that above procedure is sufficient for that. But having successful build is definitely a first step toward such goal.

Hopefully it should build also for MINGW32 environment users (assuming `mingw-w64-i686-*` packages are used, of course), but I haven't tried it.

**Comment by Lectem** [comment]
Since the build system and dependencies changed a lot since the issue was opened, I'm closing this as it may (or not) be already fixed.
If this is still an issue, please feel free to re-open it.
